---
title: "predoc Huber practical"
output:
  html_document:
    keep_md: yes
    toc: yes
    theme: united
    code_folding: hide
editor_options: 
  chunk_output_type: inline
---

Script summarising analyses of the Huber practical during the Predoc course 2022.

If you have any question just write to me :) (mira.burtscher@embl.de or on slack).

# General settings

```{r setup}
knitr::opts_chunk$set(
  echo = TRUE,
  warning = FALSE,
  message = FALSE,
  include = TRUE,
  cache = TRUE,
  cache.lazy = FALSE,
  eval = TRUE,
  fig.width = 4 * (1 + sqrt(5)) / 2,
  fig.height = 4,
  dpi = 700
)
```


## Packages

```{r, message=F, warning =F, include=F}
# general
library(tidyverse)
library(ggplot2)
theme_set(theme_bw()) # set default plot theme
library("reshape2")
library(ggrepel)
library(knitr)
library(ggpubr)
# DEG
library(DESeq2)
library(scran)
# TF activities and enrichments
library(clusterProfiler)
library(dorothea)
library(decoupleR)
options(connectionObserver = NULL)
library(org.Hs.eg.db)
# random forest
library(randomForest)
library(caret)

# set some dplyr functions as default
mutate <- dplyr::mutate
select <- dplyr::select
group_by <- dplyr::group_by
filter <- dplyr::filter
```

```{r}
options(ggplot2.discrete.colour = c("purple4", "goldenrod", "darkgreen", "darkorange", "darkblue", "darkred", "darkcyan"))
options(ggplot2.discrete.fill = c("purple4", "goldenrod", "darkgreen", "darkorange", "darkblue", "darkred", "darkcyan"))
```

```{r}
setwd("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/")
```
# Load data
All data can be downloaded here:
https://zenodo.org/record/6358987#.Y5xFqlHMLEb
Just unzip and adjust paths in the following junks.

Count data as used as input for dge

```{r}
use_sce_BR16 <-
  readRDS("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/data/sce/sce_br16.rds")
use_sce_lm2 <-
  readRDS("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/data/sce/sce_lm2.rds")
use_sce_patients <-
  readRDS("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/data/sce/sce_patient.rds")
```

Results from reproduced DGE analysis (mostly wrong because povided code does not work)

```{r}
# reproduced results = wrong
dge_BR16 <- readRDS("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/output/br16/dge_edgeR_QLF_robust_try2.rds")

# published results = correct
dge_BR16_old <- readRDS("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/output/br16/dge_edgeR_QLF_robust_OLD.rds")

# Jonas results = wrong
dge_BR16_JR <- readRDS("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/output/br16/dge_edgeR_QLF_robust_jonas.rds")

# other datasets reproduced = wrong
dge_LM2 <- readRDS("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/output/lm2/dge_edgeR_QLF_robust.rds")
dge_patient <- readRDS("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/output/patient/dge_edgeR_QLF_robust.rds")
```

# PCAs

```{r}
counts_BR16 <- use_sce_BR16@assays@data@listData[["logcpm"]]
metadata_BR16 <- colData(use_sce_BR16) %>% data.frame()

counts_lm2 <- use_sce_lm2@assays@data@listData[["logcpm"]]
metadata_lm2 <- colData(use_sce_lm2) %>% data.frame()

counts_patients <- use_sce_patients@assays@data@listData[["logcpm"]]
metadata_patients <- colData(use_sce_patients) %>% data.frame()
```


## top variable genes

### BR16

```{r}
# extract top variable genes
fit_res <- scran::modelGeneVar(use_sce_BR16, assay.type = "logcpm")
fit_md <- metadata(fit_res)
hvg_var_bio <- scran::getTopHVGs(fit_res, n = 500)
use_genes <- hvg_var_bio

m <- t(counts_BR16[match(use_genes, rownames(counts_BR16)), ])
```

Eigencorplot

```{r}
p <- PCAtools::pca(t(m), metadata = colData(use_sce_BR16) %>% data.frame())

use_metavars <- c(
  `Library size` = "sum",
  `Detected genes` = "detected",
  `Mitochondrial\nproportion` = "subsets_Mito_percent",
  `CTC type` = "sample_type",
  `Number of CTC` = "ctc_n",
  `Time point` = "timepoint",
  `Collection date` = "collection_date",
  `G1 score` = "G1_score",
  `G2M score` = "G2M_score",
  `S score` = "S_score"
)

p$metadata <- p$metadata %>%
  mutate(
    `Library size` = sum,
    `Detected genes` = detected,
    `Mitochondrial\nproportion` = subsets_Mito_percent,
    `CTC type` = sample_type,
    `Number of CTC` = ctc_n,
    `Time point` = timepoint,
    `Collection date` = collection_date,
    `G1 score` = G1_score,
    `G2M score` = G2M_score,
    `S score` = S_score
  )


PCAtools::eigencorplot(
  p,
  components = PCAtools::getComponents(p, 1:7),
  metavars = names(use_metavars),
  col = c("blue2", "blue1", "black", "red1", "red2"),
  colCorval = "white",
  scale = TRUE,
  main = "BR16 -PCs clinical correlations",
  plotRsquared = FALSE,
  signifSymbols = c("***", "**", "*", ""),
  signifCutpoints = c(0, 0.0001, 0.001, 0.01, 1)
  # cexTitleX= use_cex,
  # cexTitleY= use_cex,
  # cexLabX = use_cex,
  # cexLabY = use_cex,
  # cexMain = use_cex,
  # cexLabColKey = use_cex,
  # cexCorval = use_cex
)
```

PCA plots

```{r}
pca <- prcomp(m)
plot((pca$sdev)^2 / sum(pca$sdev^2) * 100)

# PC4 and PC5
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_BR16) %>%
  ggplot(aes(x = PC4, y = PC5, colour = timepoint, shape = sample_type)) +
  geom_point(size = 3) +
  labs(title = "BR16", subtitle = "top500")

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_BR16) %>%
  ggplot(aes(x = PC4, y = PC5, colour = timepoint, shape = collection_date)) +
  geom_point(size = 3) +
  labs(title = "BR16", subtitle = "top500")

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_BR16) %>%
  ggplot(aes(x = PC4, y = PC5, colour = cell_cycle_phase, shape = timepoint)) +
  geom_point(size = 3) +
  labs(title = "BR16", subtitle = "top500")

# PC1 and PC2
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_BR16) %>%
  ggplot(aes(x = PC1, y = PC2, colour = S_score, shape = sample_type)) +
  geom_point(size = 3) +
  labs(title = "BR16", subtitle = "top500")

pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_BR16) %>%
  ggplot(aes(x = PC1, y = PC2, colour = G2M_score, shape = sample_type)) +
  geom_point(size = 3) +
  labs(title = "BR16", subtitle = "top500")
```

### LM2

```{r}
fit_res <- scran::modelGeneVar(use_sce_lm2, assay.type = "logcpm")
fit_md <- metadata(fit_res)
hvg_var_bio <- scran::getTopHVGs(fit_res, n = 500)
use_genes <- hvg_var_bio

m <- t(counts_lm2[match(use_genes, rownames(counts_lm2)), ])
pca <- prcomp(m)
plot((pca$sdev)^2 / sum(pca$sdev^2) * 100)
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_lm2) %>%
  ggplot(aes(x = PC4, y = PC5, colour = timepoint, shape = collection_date)) +
  geom_point(size = 3) +
  labs(title = "LM2", subtitle = "top500")
```


NOT WORKING :(

```{r, eval =F}

p <- PCAtools::pca(t(m), metadata = colData(use_sce_lm2) %>% data.frame())

use_metavars <- c(
  `Library size` = "sum",
  `Detected genes` = "detected",
  `Mitochondrial\nproportion` = "subsets_Mito_percent",
  `CTC type` = "sample_type",
  `Number of CTC` = "ctc_n",
  `Time point` = "timepoint",
  `Collection date` = "collection_date",
  `G1 score` = "G1_score",
  `G2M score` = "G2M_score",
  `S score` = "S_score"
)

p$metadata <- p$metadata %>%
  mutate(
    `Library size` = sum,
    `Detected genes` = detected,
    `Mitochondrial\nproportion` = subsets_Mito_percent,
    `CTC type` = sample_type,
    `Number of CTC` = ctc_n,
    `Time point` = timepoint,
    `Collection date` = collection_date,
    `G1 score` = G1_score,
    `G2M score` = G2M_score,
    `S score` = S_score
  )


PCAtools::eigencorplot(
  p,
  metavars = colnames(p$metadata)
)
```



### patients

```{r}
fit_res <- scran::modelGeneVar(use_sce_patients, assay.type = "logcpm")
fit_md <- metadata(fit_res)
hvg_var_bio <- scran::getTopHVGs(fit_res, n = 500)
use_genes <- hvg_var_bio

m <- t(counts_patients[match(use_genes, rownames(counts_patients)), ])
pca <- prcomp(m)
plot((pca$sdev)^2 / sum(pca$sdev^2) * 100)
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_patients) %>%
  ggplot(aes(x = PC1, y = PC2, colour = timepoint, shape = collection_date)) +
  geom_point(size = 3) +
  labs(title = "patients", subtitle = "top500")
```

NOT WORKING :(


```{r, eval = F}
p <- PCAtools::pca(t(m), metadata = colData(use_sce_patients) %>% data.frame())

use_metavars <- c(
  `Library size` = "sum",
  `Detected genes` = "detected",
  `Mitochondrial\nproportion` = "subsets_Mito_percent",
  `CTC type` = "sample_type",
  `Number of CTC` = "ctc_n",
  `Time point` = "timepoint",
  `Collection date` = "collection_date",
  `G1 score` = "G1_score",
  `G2M score` = "G2M_score",
  `S score` = "S_score"
)

p$metadata <- p$metadata %>%
  mutate(
    `Library size` = sum,
    `Detected genes` = detected,
    `Mitochondrial\nproportion` = subsets_Mito_percent,
    `CTC type` = sample_type,
    `Number of CTC` = ctc_n,
    `Time point` = timepoint,
    `Collection date` = collection_date,
    `G1 score` = G1_score,
    `G2M score` = G2M_score,
    `S score` = S_score
  )


PCAtools::eigencorplot(
  p,
  components = PCAtools::getComponents(p, 1:7),
  metavars = names(use_metavars),
  col = c("blue2", "blue1", "black", "red1", "red2"),
  colCorval = "white",
  scale = TRUE,
  main = "BR16 -PCs clinical correlations",
  plotRsquared = FALSE,
  signifSymbols = c("***", "**", "*", ""),
  signifCutpoints = c(0, 0.0001, 0.001, 0.01, 1),
  corUSE = "everything"
  # cexTitleX= use_cex,
  # cexTitleY= use_cex,
  # cexLabX = use_cex,
  # cexLabY = use_cex,
  # cexMain = use_cex,
  # cexLabColKey = use_cex,
  # cexCorval = use_cex
)

PCAtools::eigencorplot(p, metavars = c("Detected genes"))
```

## all genes

Just to see how PCA looks with all genes  :)

```{r, eval = F}
m <- t(counts_BR16)
pca <- prcomp(m)
plot((pca$sdev)^2 / sum(pca$sdev^2) * 100)
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_BR16) %>%
  ggplot(aes(x = PC4, y = PC5, colour = timepoint, shape = collection_date)) +
  geom_point(size = 3) +
  labs(title = "BR16", subtitle = "all genes")


m <- t(counts_lm2)
pca <- prcomp(m)
plot((pca$sdev)^2 / sum(pca$sdev^2) * 100)
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_lm2) %>%
  ggplot(aes(x = PC1, y = PC2, colour = timepoint, shape = collection_date)) +
  geom_point(size = 3) +
  labs(title = "LM2", subtitle = "all genes")

m <- t(counts_patients)
pca <- prcomp(m)
plot((pca$sdev)^2 / sum(pca$sdev^2) * 100)
pca$x %>%
  as.data.frame() %>%
  rownames_to_column("sample_alias") %>%
  left_join(metadata_patients) %>%
  ggplot(aes(x = PC1, y = PC2, colour = timepoint, shape = collection_date)) +
  geom_point(size = 3) +
  labs(title = "Patients", subtitle = "all genes")
```


# explore dge data

## reproduced vs published

BR16 mice

```{r}
dge_BR16[["results"]] %>%
  ggplot(aes(x = logFC, y = -log10(FDR))) +
  geom_point() +
  xlim(c(-20, 20)) +
  labs(title = "Mira")

dge_BR16_old[["results"]] %>%
  ggplot(aes(x = logFC, y = -log10(FDR))) +
  geom_point() +
  labs(title = "publication")

dge_BR16_JR[["results"]] %>%
  ggplot(aes(x = logFC, y = -log10(FDR))) +
  geom_point() +
  xlim(c(-20, 20))+
  labs(title = "Jonas")
```

Other datasets 

```{r}
dge_LM2[["results"]] %>%
  ggplot(aes(x = logFC, y = -log10(FDR))) +
  geom_point() +
  xlim(c(-20, 20)) +
  labs(title = "LM2")

dge_patient[["results"]] %>%
  ggplot(aes(x = logFC, y = -log10(FDR))) +
  geom_point() +
  xlim(c(-20, 20)) +
  labs(title = "patients")
```


## DEseq2

Because theier edgeR code did not work Tuemay helped us to do the dge in DEseq. However, some filtering steps are missung.

### classic model

```{r}
# input
dds <- convertTo(use_sce_BR16, type="DESeq2", assay.type = "counts")
dds@colData$timepoint <- as.factor(dds@colData$timepoint)
dds@colData$cell_cycle_phase <- as.factor(dds@colData$cell_cycle_phase)

# model
design(dds)  <- as.formula(~ 0 + timepoint)
dds <- DESeq(dds)
resultsNames(dds)

# results
resdds_classic <- results(dds,  contrast=c("timepoint", "resting","active")) %>% 
  data.frame()

hist(resdds_classic$pvalue, breaks=100, main="RNA-Seq", xlab="p values")
```

Volcano plot

```{r}
resdds_classic %>%
  as.data.frame() %>% 
  mutate(hit = ifelse((padj < 0.05 & log2FoldChange > 0.5)|(padj < 0.05 & log2FoldChange < -0.5), "hit", "no hit")) %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj), colour = hit)) +
  geom_point() +
  labs(title ="DEseq2")
```

Convert identifiers

```{r}
library(EnsDb.Hsapiens.v79)
geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v79,
                              keys= rownames(resdds_classic),
                              keytype = "GENEID", 
                              columns = c("SYMBOL","GENEID")) %>% 
  as.data.frame()

# detach package because it is overwriting some functions
detach("package:EnsDb.Hsapiens.v79", unload = TRUE)
detach("package:ensembldb", unload = TRUE)

resdds_classic <- resdds_classic %>%
  rownames_to_column("GENEID") %>% 
  left_join(geneIDs1)
```

Comparison DEseq2 results to EdgeR results

```{r}
model_comp_classic <- dge_BR16_old[["results"]] %>%
  data.frame() %>%
  rownames_to_column("ENSEMBL") %>%
  group_by(gene_name) %>% 
  select(gene_name, logFC, FDR, description) %>% 
  inner_join(resdds_classic , by = c("gene_name"= "SYMBOL"))

ggplot(model_comp_classic,
       aes(x = logFC, y = log2FoldChange)) +
  geom_point()+
  xlim(c(-4, 4)) +
  ylim(c(-4, 4)) +
  labs(x = "EdgeR", y = "DEseq", title = "fold-change comparison")
```

### cell cycle corrected model

```{r}
# input
dds <- convertTo(use_sce_BR16, type="DESeq2", assay.type = "counts")
dds@colData$timepoint <- as.factor(dds@colData$timepoint)
dds@colData$cell_cycle_phase <- as.factor(dds@colData$cell_cycle_phase)
dds@colData$group <- as.factor(paste0(dds@colData$cell_cycle_phase, "_",dds@colData$timepoint ))

# model
design(dds)  <- as.formula(~ 0 + timepoint + cell_cycle_phase )
dds <- DESeq(dds)
resultsNames(dds)

# results
resdds_cc <- results(dds,  contrast=c("timepoint", "resting","active")) %>% 
  data.frame()

hist(resdds_cc$pvalue, breaks=100, main="RNA-Seq", xlab="p values")
```

```{r}
library(EnsDb.Hsapiens.v79)
geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v79,
  keys = rownames(resdds_cc),
  keytype = "GENEID",
  columns = c("SYMBOL", "GENEID")
) %>%
  as.data.frame()

# detach package because it is overwriting some functions
detach("package:EnsDb.Hsapiens.v79", unload = TRUE)
detach("package:ensembldb", unload = TRUE)

resdds_cc <- resdds_cc%>%
  rownames_to_column("GENEID")  %>% 
  left_join(geneIDs1)
```

Compare classic versus corrected model

```{r}
modelcomp_desdeq <- resdds_cc %>% 
  inner_join(resdds_classic , 
  by = "GENEID",
  suffix = c("_cc", "_classic"))

ggplot(modelcomp_desdeq,
       aes(x = log2FoldChange_classic, y = log2FoldChange_cc)) +
  geom_point()+
  xlim(c(-4, 4)) +
  ylim(c(-4, 4)) +
  labs(title = "model comparison")
```

## Interaction

```{r}
# input
dds <- convertTo(use_sce_BR16, type="DESeq2", assay.type = "counts")
dds@colData$timepoint <- as.factor(dds@colData$timepoint)
dds@colData$cell_cycle_phase <- as.factor(dds@colData$cell_cycle_phase)
# create grouped effect
dds@colData$group <- as.factor(paste0(dds@colData$cell_cycle_phase, "_",dds@colData$timepoint ))

design(dds)  <- as.formula(~ 0 + group )
dds <- DESeq(dds)
resultsNames(dds)

# results
resdds_cc_G2M <- results(dds,  contrast=c("group", "G2M_active","G2M_resting")) %>% 
  data.frame()

resdds_cc_S <- results(dds,  contrast=c("group", "S_active","S_resting")) %>% 
  data.frame()

hist(resdds_cc$pvalue, breaks=100, main="RNA-Seq", xlab="p values")
```

Volcano plot

```{r}
bind_rows("M-phase" = resdds_cc_G2M,
          "S-phase" = resdds_cc_S,.id = "CellCycle") %>% 
  as.data.frame() %>% 
  mutate(hit = ifelse((padj < 0.05 & log2FoldChange > 0.5)|(padj < 0.05 & log2FoldChange < -0.5), "hit", "no hit")) %>% 
  ggplot(aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(colour = hit)) +
  scale_color_manual(values = c("darkred", "grey30")) +
  facet_wrap(~CellCycle) +
  theme(legend.position = "none") +
  xlab("log2FC active-resting")
```

## Markers

Check some known markers to see if mesenchymal and epithelial phenotype is conserved.
https://www.novusbio.com/antibody-news/antibodies/antibodies-for-epithelial-mesenchymal-transition-emt-marker

```{r}
# just from the net, next time consider to ask chatGPT
markers <- data.frame(
  SYMBOL = c("CDH1", "EPCAM", "LAMB1","MUC1", "SDC1",
             "FN1", "CDH2", "CD44", "S100A4"),
  phenotype = c("epithelial", "epithelial", "epithelial", "epithelial", "epithelial", 
                "mesenchymal", "mesenchymal", "mesenchymal", "mesenchymal")
)

models_combined <- bind_rows("classic" = resdds_classic,
                             "cc_corrected" = resdds_cc,
                             .id = "model") %>% 
  inner_join(markers)

# check difference btw markers in different models
ggplot(models_combined, 
       aes(x = model, y = log2FoldChange, colour = phenotype)) +
  geom_boxplot() +
  ggforce::geom_sina(size =3) +
  #scale_color_manual(values = c("darkred", "darkgreen")) +
  labs(title = "Marker gene expression")
  
```

# Classifier

Jonas code for rf classifier

```{r}
# prepare data
# sce_BR16 = readRDS(file = file.path("./../data/sce/sce_br16.rds"))

# extract counts
BR16_rf = data.frame(t(use_sce_BR16@assays@data@listData[["logcpm"]]))

# add timepoint information
BR16_rf$timepoint = use_sce_BR16$timepoint

# convert timepoint to factor
BR16_rf$timepoint = as.factor(BR16_rf$timepoint)

# select randomly 70% of rows
set.seed(1)
BR16_train = BR16_rf[sample(nrow(BR16_rf), round(0.7*138)), ]
subset = rownames(BR16_rf) %in% rownames(BR16_train)
BR16_test = BR16_rf[subset == FALSE,]

# how balanced is the train and test set?
table(BR16_test$timepoint)
table(BR16_train$timepoint)

# train 
time_point = BR16_train$timepoint
timepoint_model = randomForest(x = dplyr::select(BR16_train, -"timepoint"), y = time_point)

# predict
set.seed(1)
predictions <- predict(timepoint_model, BR16_test)

# results
result = data.frame(timepoint = BR16_test$timepoint, prediction = predictions)
table(result$timepoint == result$prediction)
accuracy <- mean(result$timepoint == result$prediction)
accuracy

# important features for prediction
#Conditional=True, adjusts for correlations between predictors.
i_scores <- caret::varImp(timepoint_model, conditional=TRUE)
i_scores_sorted = i_scores[order(-i_scores$Overall), , drop = FALSE]

# prepare df for downstream plotting 
rf_importance <- i_scores_sorted %>% 
  rownames_to_column("GENEID") %>% 
  setNames(c("GENEID", "Overall_importance"))
```

If you do not set a seed this results will vary a lot. The current seed does not reflect the results we presented in the meeting!

## Results evaluation

This are the results generated by Jonas used for the plots of the presentation. Just play around with your own ones (or ask me/Jonas if you want the ones of Jonas) :)

```{r, eval =F}
rf_importance <- read_csv("C:/Users/burtsche/Documents/PreDoc_course/predoc_course/Huber_practical/output/i_scores.csv") %>% 
  setNames(c("GENEID", "Overall_importance"))
```

ORA enrichment

```{r}
i_scores_subset <- rf_importance %>% 
  filter(Overall_importance > 0.05)

ER_GO <- enrichGO(
  gene = i_scores_subset$GENEID,
  keyType = "ENSEMBL",
  OrgDb = "org.Hs.eg.db",
  pvalueCutoff = 1, 
  ont = "ALL"
) %>%
  as.data.frame()

# plot GO
ggplot(
  ER_GO %>%
    arrange(qvalue) %>%
    slice_head(n = 15),
  aes(
    x = -log10(qvalue),
    y = reorder(Description, -log10(qvalue)),
    size = Count,
    colour = -log10(qvalue)
  )
) +
  geom_point() +
  # facet_wrap(~ONTOLOGY, scales = "free", ncol  =1) +
  labs(y = "")
```

Hit comparison

```{r}
hits_classic <- resdds_classic %>% 
  filter(padj < 0.05)

hits_cc <- resdds_cc %>% 
  filter(padj < 0.05)

hits_rf <- rf_importance %>%
  ungroup() %>% 
  filter(Overall_importance > 0.05)

method_comparison <- 
  data.frame(
    GENEID = unique(c(hits_cc$GENEID, hits_rf$GENEID, hits_classic$GENEID))) %>% 
  mutate(
    rf = ifelse(GENEID %in% hits_rf$GENEID, TRUE, FALSE),
    cc = ifelse(GENEID %in% hits_cc$GENEID, TRUE, FALSE),
    classic = ifelse(GENEID %in% hits_classic$GENEID, TRUE, FALSE)
  )
```

```{r}
method_comparison %>%  
ggplot() +
  ggvenn::geom_venn(aes(
    A = `rf`,
    B = `cc`,
    C = `classic`
  ),
  fill_color = c("darkorange", "darkblue", "darkviolet"),
  fill_alpha = 1,
  stroke_alpha = 1,
  stroke_color = "white",
  set_name_size = 5,
  set_names = c("random forest", "cc corrected", "classic"),
  set_name_color = c("darkorange", "darkblue", "darkviolet"),
  text_size = 4,
  text_color = "white",
  show_percentage = FALSE
  ) +
  coord_fixed() +
  theme_void() 
```


# TF activity
## database

```{r}
# load database
# attention: it is a human xenograft!! so I should use the human database

dorothea_df <- dorothea_hs %>%
  # only high confidence levels
  dplyr::filter(confidence %in% c("A", "B", "C")) %>%
  dplyr::select(target, tf, mor) %>%
  as.data.frame() %>%
  mutate(target = toupper(target), tf = toupper(tf))

dorothea_df$likelihood <- 1
```

## input

```{r}
# prepare input
# use only genes characteristic for either active or resting

input_day <- dge_BR16_old[["results"]] %>%
  data.frame() %>%
  rownames_to_column("ENSEMBL") %>%
  group_by(gene_name) %>%
  summarise(logFC = mean(logFC)) %>%
  # early timepoint
  filter(logFC >= 0.5) %>%
  column_to_rownames("gene_name")

input_night <- dge_BR16_old[["results"]] %>%
  data.frame() %>%
  rownames_to_column("ENSEMBL") %>%
  group_by(gene_name) %>%
  summarise(logFC = mean(logFC)) %>%
  # late timepoint
  filter(logFC <= -0.5) %>%
  column_to_rownames("gene_name")
```

## run decupleR

https://github.com/saezlab/decoupleR
Tool which collects a series of methods to get TF activities. 

```{r}
TF_activities_day <- as.data.frame(
  run_viper(
    mat = as.matrix(input_day),
    network = dorothea_df,
    .source = "tf"
  )
)


TF_activities_night <- as.data.frame(
  run_viper(
    mat = as.matrix(input_night),
    network = dorothea_df,
    .source = "tf"
  )
)
```

Plot and compare results

```{r}
# compare scores
full_join(TF_activities_day, TF_activities_night,
  by = c("source", "condition", "statistic"),
  suffix = c("_day", "_night")
) %>%
  ggplot(aes(x = score_day, y = score_night, label = source)) +
  geom_abline(slope = 1) +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0) +
  geom_point() +
  ggrepel::geom_text_repel() +
  xlim(c(-4, 4)) +
  ylim(c(-4, 4))

# top hits
bind_rows("active" = TF_activities_day,
          "resting" = TF_activities_night,
          .id = "timepoint") %>% 
  group_by(timepoint) %>% 
  slice_max(n = 5, abs(score)) %>% 
  ggplot(aes(x = score, y = reorder(source, score), fill = timepoint)) +
  geom_bar(stat = "identity") +
  facet_wrap(~timepoint, scales = "free", ncol = 1) +
  labs(y = "Transcription factor")

# ratio plot
inner_join(TF_activities_day, TF_activities_night,
  by = c("source", "condition", "statistic"),
  suffix = c("_active", "_resting")
) %>%
  filter(p_value_resting < 0.1 | p_value_active < 0.1) %>% 
  mutate(ratio = score_active / score_resting) %>%
  ggplot(aes(x = reorder(source, ratio), y = score_active / score_resting)) +
  geom_point(size=3) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)) +
  xlab("Transcription factor")
```

### example 1: MYC

```{r}
t <- dorothea_df %>%
  filter(tf == "MYC")

dge_BR16_old[["results"]] %>%
  mutate(
    label = ifelse(gene_name %in% t$target, gene_name, NA),
    MYC_targets = ifelse(gene_name %in% t$target, "MYC_targets", "")
  ) %>%
  filter(MYC_targets == "MYC_targets") %>%
  ggplot(aes(x = logFC, y = -log10(FDR), label = label, colour = MYC_targets)) +
  geom_point() +
  geom_text_repel()+
  labs(title = "MYC target expression")
```

### example 2: TEAD1

```{r}
t <- dorothea_df %>%
  filter(tf == "TEAD1")

dge_BR16_old[["results"]] %>%
  mutate(
    label = ifelse(gene_name %in% t$target, gene_name, NA),
    TEAD_targets = ifelse(gene_name %in% t$target, "TEAD_targets", "")
  ) %>%
  filter(TEAD_targets == "TEAD_targets") %>%
  ggplot(aes(x = logFC, y = -log10(FDR), label = label)) +
  geom_point() +
  geom_text_repel(size =1) +
  labs(title = "TEAD1 target expression")
```


### ORA

Over-representation analysis
Old. Very few genes!


```{r, eval = FALSE}
ip_day <- TF_activities_day %>%
  filter(p_value < 0.05)

ER_GO_day <- enrichGO(
  gene = ip_day$source,
  keyType = "SYMBOL",
  OrgDb = "org.Hs.eg.db",
  pvalueCutoff = 1, ont = "ALL"
) %>%
  as.data.frame()

ip_night <- TF_activities_night %>%
  filter(p_value < 0.05)

ER_GO_night <- enrichGO(
  gene = ip_night$source,
  keyType = "SYMBOL",
  OrgDb = "org.Hs.eg.db",
  pvalueCutoff = 1, ont = "ALL"
) %>%
  as.data.frame()

ER_GO <- enrichGO(
  gene = unique(c(ip_night$source, ip_day$source)),
  keyType = "SYMBOL",
  OrgDb = "org.Hs.eg.db",
  pvalueCutoff = 1, ont = "ALL"
) %>%
  as.data.frame()
```


### GSEA

Gene set enrichment analysis
Old. just try to see if there is something, but not really

```{r, eval = F}
ip <- TF_activities_night$score
names(ip) <- TF_activities_night$source
ip <- sort(ip, decreasing = TRUE)

set.seed(1)
ER <- gseGO(
  geneList = ip,
  keyType = "SYMBOL",
  ont = "ALL",
  OrgDb = "org.Hs.eg.db",
  pvalueCutoff = 1,
  seed = TRUE
) %>%
  as.data.frame() %>%
  mutate(length = nrow(ip))

rm(ip, input, mapping)
```

# End

```{r}
styler:::style_active_file()
```

```{r}
```
